Ghidra/Debug/Debugger-agent-gdb/src/main/java/agent/gdb/manager/parsing/GdbMiParser.java

```diff
@@ -354,40 +352,75 @@ public Object parseObject() throws GdbParseError {
 	}
 
 	/**
-	 * Parse the string at the cursor
+	 * Parse the string at the cursor, undoing GDB's printchar transformation.
 	 * 
 	 * @see #parseString(CharSequence)
 	 * @return the string
 	 * @throws GdbParseError if no text matches the pattern
 	 */
 	public String parseString() throws GdbParseError {
-		/*
-		 * Matching CSTRING for inputs of too many characters (2048, really?) causes a
-		 * StackOverflowException in Java's built-in Pattern object. Boo! Thus, I'll write this
-		 * myself. All said and done, this might actually look better than the old regex
-		 */
-		// String match = match(CSTRING);
-		//return StringEscapeUtils.unescapeJava(match.substring(1, match.length() - 1));
-		int start = buf.position();
 		if ('"' != peek(false)) { // Keep whitespace that is in the string
 			throw new GdbParseError("\"", buf);
 		}
 		buf.get(); // consume "
+
+		ByteArrayOutputStream baos = new ByteArrayOutputStream();
 		while (true) {
-			char c = buf.get();
-			if (c == '"') {
+			char ch = buf.get();
+			if (ch > 0xff) {
+				throw new GdbParseError("byte", "U+" + String.format("%04X", ch));
+			} else if (ch == '"') {
 				break;
+			} else if (ch != '\\') {
+				baos.write(ch);
+				continue;
 			}
-			else if (c == '\\') {
-				buf.get();
+
+			/* Handle escape */
+			ch = buf.get();
+			switch (ch) {
+				case 'n':
+					baos.write('\n');
+					break;
+				case 'b':
+					baos.write('\b');
+					break;
+				case 't':
+					baos.write('\t');
+					break;
+				case 'f':
+					baos.write('\f');
+					break;
+				case 'r':
+					baos.write('\r');
+					break;
+				case 'e':
+					baos.write(0x1b);
+					break;
+				case 'a':
+					baos.write(0x07);
+					break;
+				case '0':
+				case '1':
+				case '2':
+				case '3':
+					char ch2 = buf.get();
+					if (ch2 < '0' || ch2 > '9') {
+						throw new GdbParseError("octal", "\\" + ch + ch2);
+					}
+					char ch3 = buf.get();
+					if (ch3 < '0' || ch3 > '9') {
+						throw new GdbParseError("octal", "\\" + ch + ch2 + ch3);
+					}
+					int octchar = ((ch - '0') << 6) | ((ch2 - '0') << 3) | (ch3 - '0');
+					baos.write(octchar);
+					break;
+				default:
```

Comment:
If we only consider what GDB will produce, vs. what it will accept, then true, we should not expect to see a first digit above 3. I do wonder if we should explicitly forbid, e.g., `\4`, since we should never expect GDB to produce that (though it will accept it and echo back `\004`).

Reply:
I suppose if we got pedantic, we might also want to throw an exception on e.g., `\c`, because I don't know any case where GDB would produce that. It would just print `c`. I tend to prefer fairly strict checking, though that's bitten me in the past.... I wonder if the default case should just throw an exception, because we shouldn't expect GDB to use unnecessary escaping.